docs/examples/use_cases/tensorflow/efficientdet/pipeline/dali/ops.py

```diff
@@ -0,0 +1,216 @@
+    images = dali.fn.decoders.image(
+        inputs["image/encoded"],
+        device="mixed" if device == "gpu" else "cpu",
+        output_type=dali.types.RGB,
+    )
+    xmin = inputs["image/object/bbox/xmin"]
+    xmax = inputs["image/object/bbox/xmax"]
+    ymin = inputs["image/object/bbox/ymin"]
+    ymax = inputs["image/object/bbox/ymax"]
+    bboxes = dali.fn.transpose(
+        dali.fn.stack(xmin, ymin, xmax, ymax), perm=[1, 0], device="cpu"
+    )
+    classes = dali.fn.cast(
+        inputs["image/object/class/label"], dtype=dali.types.INT32, device="cpu"
+    )
+    return (
+        images,
+        bboxes,
+        classes,
+        dali.fn.cast(inputs["image/width"], dtype=dali.types.FLOAT),
+        dali.fn.cast(inputs["image/height"], dtype=dali.types.FLOAT),
+    )
+
+
+def input_coco(
+        images_path, annotations_path, device, shard_id, num_shards, random_shuffle=True
+):
+    encoded, bboxes, classes = dali.fn.readers.coco(
+        file_root=images_path,
+        annotations_file=annotations_path,
+        ratio=True,
+        ltrb=True,
+        shard_id=shard_id,
+        num_shards=num_shards,
+        random_shuffle=random_shuffle)   
+
+
+    images = dali.fn.decoders.image(
+        encoded,
+        device="mixed" if device == "gpu" else "cpu",
+        output_type=dali.types.RGB,
+    )
+
+    shape = dali.fn.peek_image_shape(encoded)
+    h = dali.fn.slice(shape, 0, 1, axes=(0,))
+    w = dali.fn.slice(shape, 1, 1, axes=(0,))
+
+    return (
+        images,
+        bboxes,
+        classes,
+        dali.fn.cast(w, dtype=dali.types.FLOAT),
+        dali.fn.cast(h, dtype=dali.types.FLOAT),
+    )
+
+
+def normalize_flip(device, images, bboxes, p=0.5):
+    flip = dali.fn.random.coin_flip(probability=p, device="cpu")
+    images = dali.fn.crop_mirror_normalize(
+        images,
+        mirror=flip.gpu() if device == "gpu" else flip,
+        mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
+        std=[0.229 * 255, 0.224 * 255, 0.225 * 255],
+        output_layout=dali.types.NHWC,
+        device=device,
+    )
+    bboxes = dali.fn.bb_flip(bboxes, horizontal=flip, ltrb=True, device="cpu")
+    return images, bboxes
+
+
+def gridmask(device, images, widths, heights):
+    if device == "gpu":
+        widths = widths.gpu()
+        heights = heights.gpu()
+
+    ratio = 0.4
+    angle = (
+        dali.fn.random.normal(mean=-1, stddev=1, device=device)
+        * 10.0
+        * (math.pi / 180.0)
+    )
+
+    l = dali.math.min(0.5 * heights, 0.3 * widths, device=device)
+    r = dali.math.max(0.5 * heights, 0.3 * widths, device=device)
+    tile = dali.fn.cast(
+        (dali.fn.random.uniform(range=[0.0, 1.0], device=device) * (r - l) + l),
+        dtype=dali.types.INT32,
+        device=device,
+    )
+
+    gridmask = dali.fn.grid_mask(
+        images, ratio=ratio, angle=angle, tile=tile, device=device
+    )
+
+    p = dali.fn.random.coin_flip(device=device)
+    images = images * (1 - p) + gridmask * p
```

Comment:
Perhaps instead of masking entire image (a lot of memory bandwidth) just mask `ratio`? With ratio being 0 (or 1, I don't remember), gridmask should be a no-op.

Reply:
Changed.